bin/node-template/Cargo.toml

```diff
@@ -18,25 +18,25 @@ tokio = "0.1.22"
 parking_lot = "0.9.0"
 codec = { package = "parity-scale-codec", version = "1.0.0" }
 trie-root = "0.15.2"
-sp-io = { path = "../../primitives/io" }
-sc-cli = { path = "../../client/cli" }
-sp-core = { path = "../../primitives/core" }
-sc-executor = { path = "../../client/executor" }
-sc-service = { path = "../../client/service" }
-sp-inherents = { path = "../../primitives/inherents" }
-sc-transaction-pool = { path = "../../client/transaction-pool" }
-sp-transaction-pool = { path = "../../primitives/transaction-pool" }
-sc-network = { path = "../../client/network" }
-sc-consensus-aura = { path = "../../client/consensus/aura" }
-sp-consensus-aura = { path = "../../primitives/consensus/aura" }
-sp-consensus = { path = "../../primitives/consensus/common" }
-grandpa = { package = "sc-finality-grandpa", path = "../../client/finality-grandpa" }
-grandpa-primitives = { package = "sp-finality-grandpa", path = "../../primitives/finality-grandpa" }
-sc-client = {  path = "../../client/" }
-node-template-runtime = { path = "runtime" }
-sp-runtime = { path = "../../primitives/runtime" }
-sc-basic-authority = { path  = "../../client/basic-authorship"}
+sp-io = { version = "2.0.0", path = "../../primitives/io" }
```

Comment:
why set versions as well as path? it's just redundant additional tedious data to maintain isn't it?

Reply:
Unfortunately, defining the version is a requirement to publish on `crates.io`. Cargo doesn't have any `publish --all` and won't resolve path to figure out which version to us–in general its publishing helper for bigger workspaces are pretty rudimentary. On the other side `path` is obviously a lot easier and nicer for development.

Fortunately, because cargo uses semver this isn't much of a deal. We can just define `version = "2.0.0"` and even if we bump the minor or patch version of that crate to make a new release, this doesn't need any update on this crate, but cargo will figure out the best version on `cargo update` (or `cargo build` if first time). Our internal `Cargo.lock` will also update without problems. This also allows us to do version updates more independently per crate, like updating network and consensus and apply the changes to the crates that follow without having to publish a breaking change, if those dependencies don't alter their externally perceived API.

Having both is the common way to stay sane, and is what we do in the [libp2p workspace repo](https://github.com/libp2p/rust-libp2p/blob/1aeb264cbc0e262af219c92e18eb0025724a9005/transports/dns/Cargo.toml#L13) and common practice on other workspace repos, too.

Reply:
👍